### ISSUE TITLE
Custom loss function from python script 

### ISSUE THREAD

**Issue by Cdk29** [rstudio/keras#1166]
Hi.

Some recents tricks are sometimes only wrote in python. One recent case I have faced is the will to use BiTempered LogisticLoss,  a different loss function, to train a model with a dataset with noisy labels. The code can be found [here](https://github.com/google/bi-tempered-loss) or [here](https://www.kaggle.com/durbin164/bitempered-logistic-loss-tensorflow-v2?select=bi_tempered_loss.py).

I have tried to wrote the following :

```
library(reticulate)

source_python("../input/bitempered-loss/loss.py")

bi_tempered <- function(y_true, y_pred) {
    
  bi_tempered_logistic_loss(activations=y_true, labels=y_pred, t1=0.6, t2=1.4, label_smoothing=0.1, num_iters=5L)

}
```

I did not managed to got it work, but it is not the point of this issue. Debugging with reticulate can be quite tricky, so I was not fully sure about what caused the issue. The tensor that I got back from the loss function did not seemed correct, but I will not bet on it.

I would like to know if passing a loss function from a python script directly via reticulate is indeed possible, or if I tried to do something impossible. Especially, I had to write **num_iters=5L** (python did not understood the 5 as an int otherwise), so I wonder if the tensor that R passed to Python was something Python could handle, and vice versa, if what python gave back was something that the keras interface could handle.

To rephrase it : the [website of reticulate](https://rstudio.github.io/reticulate/), as a list of _Type conversions_, but tensor are not listed in it. Can R pass tensor from the wrapper to Python and vice versa ?

**Comment by dfalbel** [comment]
Hi @Cdk29 

AFAICT this is possible. For example, this works as expected:

``` r
library(keras)

define_model <- function() {
  model <- keras_model_sequential() 
  model %>%
    layer_dense(32, input_shape = 784, kernel_initializer = initializer_ones()) %>%
    layer_activation('relu') %>%
    layer_dense(10) %>%
    layer_activation('softmax')
  model
}

model <- define_model()
pyfun <- reticulate::py_run_string("
import tensorflow as tf
def loss_fn (y_true, y_pred):
  return tf.keras.losses.categorical_crossentropy(y_true, y_pred)

")

model %>% 
  compile(
    loss = pyfun$loss_fn,
    optimizer = "adam"
  )

# generate dummy training data
data <- matrix(rexp(1000*784), nrow = 1000, ncol = 784)
labels <- matrix(round(runif(1000*10, min = 0, max = 9)), nrow = 1000, ncol = 10)


model %>% fit(x = data, y = labels)
```

<sup>Created on 2021-02-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
